src/pydicom/filewriter.py

```diff
@@ -901,80 +937,170 @@ def write_file_meta_info(
-    #   requires Explicit VR Little Endian) and COMMAND SET (which requires
-    #   Implicit VR Little Endian)
-    fp.is_implicit_VR = cast(bool, dataset.is_implicit_VR)
-    fp.is_little_endian = cast(bool, dataset.is_little_endian)
-
-    # Write non-Command Set elements now
-    write_dataset(fp, get_item(dataset, slice(0x00010000, None)))
+        return cast(tuple[bool, bool], arg_encoding)
+
+    # The default for little_endian is `None` so we can require the use of
+    #   args with `force_encoding`, but we actually default it to `True`
+    #   when `implicit_vr` is used as a fallback
+    if implicit_vr is not None and little_endian is None:
+        arg_encoding = (implicit_vr, True)
+
+    ds_encoding: EncodingType = (None, None)
+    if not config._use_future:
+        ds_encoding = (ds.is_implicit_VR, ds.is_little_endian)
+
+    fallback_encoding: EncodingType = (None, None)
+    if None not in arg_encoding:
+        fallback_encoding = arg_encoding
+    elif None not in ds_encoding:
+        fallback_encoding = ds_encoding
+    elif None not in ds.original_encoding:
+        fallback_encoding = ds.original_encoding
+
+    if tsyntax is None:
+        if None not in fallback_encoding:
+            return cast(tuple[bool, bool], fallback_encoding)
+
+        raise ValueError(
+            "Unable to determine the encoding to use for writing the dataset, "
+            "please set the file meta's Transfer Syntax UID or use the "
+            "'implicit_vr' and 'little_endian' arguments"
+        )
+
+    if tsyntax.is_private and not tsyntax.is_transfer_syntax:
+        if None in fallback_encoding:
+            raise ValueError(
+                "The 'implicit_vr' and 'little_endian' arguments are required "
+                "when using a private transfer syntax"
+            )
+
+        return cast(tuple[bool, bool], fallback_encoding)
+
+    if not tsyntax.is_transfer_syntax:
+        raise ValueError(
+            f"The Transfer Syntax UID '{tsyntax.name}' is not a valid "
+            "transfer syntax"
+        )
+
+    # Check that supplied args match transfer syntax
+    if implicit_vr is not None and implicit_vr != tsyntax.is_implicit_VR:
+        raise ValueError(
+            f"The 'implicit_vr' value is not consistent with the required "
+            f"VR encoding for the '{tsyntax.name}' transfer syntax"
+        )
+
+    if little_endian is not None and little_endian != tsyntax.is_little_endian:
+        raise ValueError(
+            f"The 'little_endian' value is not consistent with the required "
+            f"endianness for the '{tsyntax.name}' transfer syntax"
+        )
+
+    return (tsyntax.is_implicit_VR, tsyntax.is_little_endian)
 
 
 def dcmwrite(
     filename: PathType | BinaryIO | DicomFileLike,
     dataset: Dataset,
-    write_like_original: bool = True,
+    __write_like_original: bool = True,
+    implicit_vr: bool | None = None,
```

Comment:
I wanted to keep backwards compatibility for argument order and still be able to test if the `write_like_original` kwarg was used.

ie: `dcmwrite(ds, False, implicit_vr=True)` and `dcmwrite(ds, write_like_original=False, implicit_vr=True)` should be backwards compatible and equivalent but still flag the deprecation.

Reply:
Ah, I see, basically an argument placeholder. I expect people always used the keyword, because that was the easy it was used in examples and discussions. So, maybe could make all the arguments except dataset keyword-only with a `*,` parameter?  Technically backwards incompatible, but probably rarely hit and pretty clear when the error is thrown. Then the arg could be removed entirely in v4.

Reply:
OK, sounds reasonable

Reply:
OK, on trying to implement - the default is `True`, so the argument will always be used, so I'll always be flagging the deprecation no matter if the users actually use `write_like_original` or not.

Reply:
I can only issue deprecation/`use_future` exceptions for `write_like_original=False` though.

Reply:
I've played with something that I think captures all cases.  Basically it is the same as you have, except the positional argument(s) is/are captured with a preceding `*`.  I printed "Deprecation warning" because Python doesn't issue the same one twice.  This allows a position argument, but gives warning if it is used (or SyntaxError if `_use_future` on).  In v4, we could keep `*_pos_args` with our own error message or replace it with a `*, ` to force a Syntax error if any positional arg passed.

```python
import warnings
from pydicom import config


def dcmwrite(filename, dataset, *_pos_args, 
            implicit_vr=None, little_endian=None, enforce_file_format=False,
            force_encoding=False, **kwargs
    ):
    print(f"{filename=}", f"{dataset=}", f"{_pos_args=}", f"{kwargs=}")

    write_like_val = None

    if _pos_args:
        # TODO: also error on len(write_like_original) > 1
        write_like_val = _pos_args[0]
    elif "write_like_original" in kwargs:
        write_like_val = kwargs["write_like_original"]

    if write_like_val is not None:
        msg = (
            "`write_like_original` is no longer accepted as a positional or keyword argument. "
            f"Please use keyword argument `enforce_file_format={write_like_val}`"
        )

        if config._use_future:
            raise SyntaxError(msg)
        else:
            print("Deprecation warning")
            warnings.warn(msg, DeprecationWarning)


# config._use_future = True

print("\njust filename, dataset")
dcmwrite("testfile", "dataset")

print("\nwrite_like_original supplied positional:")
dcmwrite("testfile", "dataset", True, implicit_vr=True)

print("\nwrite_like_original supplied keyword")
dcmwrite("testfile", "dataset", write_like_original=True, implicit_vr=True)
```

Reply:
This seems really similar to what I had? Should I just revert it to what it was before?

Reply:
```python
def dcmwrite(
    filename: PathType | BinaryIO | DicomFileLike,
    dataset: Dataset,
    *args,
    implicit_vr: bool | None = None,
    little_endian: bool | None = None,
    enforce_file_format: bool = False,
    force_encoding: bool = False,
    **kwargs,
) -> None:
    if args and len(args) != 1:
        raise TypeError(
            f"dcmwrite() takes at most 3 positional arguments ({len(args) + 2} "
            "given)"
        )

    write_like_original = kwargs.get("write_like_original", None)
    if args and write_like_original is not None:
        raise TypeError(
            "'write_like_original' used as both a positional and keyword argument"
        )

    # Use args only if no `write_like_original`
    if args and write_like_original is not None:
        write_like_original = args[0]

    if write_like_original is not None:
        if config._use_future:
            raise TypeError(
                "Invalid keyword argument for dcmwrite(): 'write_like_original'"
            )

        warnings.warn(
            (
                "'write_like_original' is deprecated and will be removed in "
                "v4.0, please use 'enforce_file_format="
                f"{not write_like_original}' instead"
            ),
            DeprecationWarning,
        )
        enforce_file_format = not write_like_original
```
How's this? I honestly think my original version was simpler, though.

Reply:
> How's this? I honestly think my [original version](https://github.com/scaramallion/pydicom/blob/e90d732955eb49d8c4e94f70b24ba86a12efa556/src/pydicom/filewriter.py#L1220) was simpler, though.

Apologies for dragging this out ...

I didn't explain myself well.  The key thing I was thinking about was what happens in the distant future, with current user code like: `dcmwrite(fp, ds, False)`.  At some point I'm assuming the fake `__write_like_original` positional argument would be removed, then that user code line would lead the `False` to be assigned to the first new argument instead, a very tricky hidden error.  So the key thing that was needed to prevent that is to force the new arguments to be keyword-only.  That could be via `*args`, or (now that I think about it) perhaps even better with just `(...,__write_like_original:bool = True, *, ...)`.  The new pydicom code around this change has *used* them as keyword-only, but to me it needs to be forced by the function signature for user code.

The second key thing I was trying to do is signal the future 'breakage' of `write_like_original` as a positional argument in v3, in prep for that being an error in v4, and either then or at some point a syntax error when `__write_like_original` is removed.

If the code can be set to handle those two principles, then I'm not particularly concerned about the exact format.  I don't think the current or previous one handle those exactly, unless I'm misunderstanding something...

> I can only issue deprecation warnings/`use_future` exceptions for `write_like_original=False` though.

And I'm not sure I understood the reason for that statement, but the code I proposed does signal future error/warning for any value of `write_like_original`.

Reply:
OK, fair enough

Reply:
Alright, I think I've covered the behaviour you want. Using `write_like_original` as either positional or keyword will always flag a deprecation warning or `TypeError` if `_use_future`. Feel free to nitpick though, I'd rather get it right.

I've also simplified `save_as()` to be more of a pass-through to `dcmwrite()`